File: backend/app/api/routes/benchmark_review.py
from uuid import UUID

from fastapi import APIRouter, Query

from app.api.dependencies import CurrentAdmin, CurrentSuperAdmin, DbSession
from app.schemas.benchmark_review import (
    BenchmarkAuditRead, BenchmarkCaseRead, BenchmarkCriteriaWrite, BenchmarkDetail, BenchmarkExport,
    BenchmarkImportWrite, BenchmarkPublicationRead, BenchmarkPublishWrite, BenchmarkReviewImportWrite,
    BenchmarkReviewWrite, BenchmarkSummary,
)
from app.services import benchmark_review_service as service

router = APIRouter()


@router.get("", response_model=list[BenchmarkSummary])
def list_benchmarks(actor: CurrentAdmin, db: DbSession):
    return service.list_benchmarks(db)


@router.post("", response_model=BenchmarkDetail, status_code=201)
def import_benchmark(payload: BenchmarkImportWrite, actor: CurrentSuperAdmin, db: DbSession):
    return service.import_benchmark(db, payload, actor)


@router.get("/{benchmark_id}", response_model=BenchmarkDetail)
def read_benchmark(benchmark_id: UUID, actor: CurrentAdmin, db: DbSession):
    return service.detail(db, benchmark_id)


@router.get("/{benchmark_id}/cases/{case_id}", response_model=BenchmarkCaseRead)
def read_case(benchmark_id: UUID, case_id: str, actor: CurrentAdmin, db: DbSession,
              offset: int = Query(0, ge=0)):
    return service.case_detail(db, benchmark_id, case_id, offset=offset)


@router.post("/{benchmark_id}/cases/{case_id}", response_model=BenchmarkCaseRead, status_code=201)
def save_review(benchmark_id: UUID, case_id: str, payload: BenchmarkReviewWrite, actor: CurrentAdmin, db: DbSession):
    return service.save_review(db, benchmark_id, case_id, payload, actor)


@router.post("/{benchmark_id}/criteria", response_model=BenchmarkDetail, status_code=201)
def save_criteria(benchmark_id: UUID, payload: BenchmarkCriteriaWrite, actor: CurrentSuperAdmin, db: DbSession):
    return service.save_criteria(db, benchmark_id, payload, actor)


@router.post("/{benchmark_id}/publications", response_model=BenchmarkPublicationRead, status_code=201)
def publish(benchmark_id: UUID, payload: BenchmarkPublishWrite, actor: CurrentSuperAdmin, db: DbSession):
    return service.publish(db, benchmark_id, payload, actor)


@router.get("/{benchmark_id}/export", response_model=BenchmarkExport)
def export(benchmark_id: UUID, actor: CurrentAdmin, db: DbSession,
           publication: int | None = Query(None, ge=1)):
    return service.export_bundle(db, benchmark_id, publication)


@router.post("/{benchmark_id}/review-imports", response_model=BenchmarkDetail, status_code=201)
def import_reviews(benchmark_id: UUID, payload: BenchmarkReviewImportWrite, actor: CurrentSuperAdmin, db: DbSession):
    return service.import_reviews(db, benchmark_id, payload, actor)


@router.get("/{benchmark_id}/history", response_model=BenchmarkAuditRead)
def history(benchmark_id: UUID, actor: CurrentAdmin, db: DbSession, offset: int = Query(0, ge=0)):
    return service.audit(db, benchmark_id, offset=offset)

File: frontend/src/api/benchmarkReview.ts
import { apiRequest } from "./client";
import type { components } from "../types/api.generated";

export type BenchmarkSummary = components["schemas"]["BenchmarkSummary"];
export type BenchmarkDetail = components["schemas"]["BenchmarkDetail"];
export type BenchmarkCase = components["schemas"]["BenchmarkCaseRead"];
export type ReviewWrite = components["schemas"]["BenchmarkReviewWrite"];
export type CriteriaWrite = components["schemas"]["BenchmarkCriteriaWrite"];
export type BenchmarkAudit = components["schemas"]["BenchmarkAuditRead"];
export type BenchmarkExport = components["schemas"]["BenchmarkExport"];
export type ReviewState = components["schemas"]["BenchmarkCaseSummary"]["state"];

const base = "/admin/research-quality/benchmarks";
const casePath = (id: string, caseId: string) => `${base}/${id}/cases/${encodeURIComponent(caseId)}`;
export const benchmarkApi = {
  list: () => apiRequest<BenchmarkSummary[]>(base),
  detail: (id: string) => apiRequest<BenchmarkDetail>(`${base}/${id}`),
  case: (id: string, caseId: string, offset = 0) => apiRequest<BenchmarkCase>(`${casePath(id, caseId)}?offset=${offset}`),
  saveReview: (id: string, caseId: string, body: ReviewWrite) => apiRequest<BenchmarkCase>(casePath(id, caseId), { method: "POST", body }),
  saveCriteria: (id: string, body: CriteriaWrite) => apiRequest<BenchmarkDetail>(`${base}/${id}/criteria`, { method: "POST", body }),
  publish: (id: string, expected_revision: number, reason: string) =>
    apiRequest<components["schemas"]["BenchmarkPublicationRead"]>(`${base}/${id}/publications`, { method: "POST", body: { expected_revision, reason } }),
  history: (id: string, offset = 0) => apiRequest<BenchmarkAudit>(`${base}/${id}/history?offset=${offset}`),
  export: (id: string, publication?: number) => apiRequest<BenchmarkExport>(`${base}/${id}/export${publication ? `?publication=${publication}` : ""}`),
  importBenchmark: (benchmark: unknown) => apiRequest<BenchmarkDetail>(base, { method: "POST", body: { benchmark, permissions_checked: true } }),
  importReviews: (id: string, expected_revision: number, reviews: unknown) =>
    apiRequest<BenchmarkDetail>(`${base}/${id}/review-imports`, { method: "POST", body: { expected_revision, reviews } }),
};

export function downloadJson(name: string, value: unknown) {
  const url = URL.createObjectURL(new Blob([JSON.stringify(value, null, 2)], { type: "application/json" }));
  const link = document.createElement("a");
  link.href = url; link.download = name; link.click();
  setTimeout(() => URL.revokeObjectURL(url), 1000);
}
